lib/Sema/CSGen.cpp

```diff
@@ -1539,23 +1551,51 @@ namespace {
             CS.getConstraintLocator(expr),
             {expr->getArgumentLabels(),
              expr->getUnlabeledTrailingClosureIndex()});
-        return baseTy;
+        return outputTy;
       }
-
-      // Otherwise, the member needs to be convertible to the base type.
-      CS.addConstraint(ConstraintKind::Conversion, memberTy, baseTy,
-        CS.getConstraintLocator(expr, ConstraintLocator::RValueAdjustment));
       
-      // The member type also needs to be convertible to the context type, which
-      // preserves lvalue-ness.
-      auto resultTy = CS.createTypeVariable(memberLocator,
-                                            TVO_CanBindToLValue |
-                                            TVO_CanBindToNoEscape);
-      CS.addConstraint(ConstraintKind::Conversion, memberTy, resultTy,
-                       memberLocator);
-      CS.addConstraint(ConstraintKind::Equal, resultTy, baseTy,
-                       memberLocator);
-      return resultTy;
+      // Otherwise, the result is just the type of the member.
+      return memberTy;
+    }
+
+    Type visitUnresolvedMemberChainResultExpr(
+        UnresolvedMemberChainResultExpr *expr) {
+      auto *tail = expr->getSubExpr();
+      assert(isa<UnresolvedMemberExpr>(tail) || isa<UnresolvedDotExpr>(tail) ||
+             isa<CallExpr>(tail) || isa<BindOptionalExpr>(tail) ||
+             isa<ForceValueExpr>(tail) ||
+             isa<SubscriptExpr>(tail) &&
+                 "Unexpected expression at end of unresolved member chain");
+
+      auto memberTy = CS.getType(tail);
+      auto *base = TypeChecker::getUnresolvedMemberChainBase(tail);
```

Comment:
Can we assert that `base` isn't `nullptr` instead of asserting that `tail` is of a given kind?

Reply:
With changes made re: Pavel's comment above, I've asserted that we have a base on construction of `UnresolvedMemberChainResultExpr`, and changed the assertion here to check that `base == TypeChecker::getUnresolvedMemberChainBase(tail)`.